src/lender/tranche/operator/proportional.sol

```diff
@@ -0,0 +1,173 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.15 <0.6.0;
+
+import "ds-note/note.sol";
+import "tinlake-math/math.sol";
+import "tinlake-auth/auth.sol";
+
+contract TrancheLike {
+    function supply(address usr, uint currencyAmount, uint tokenAmount) public;
+    function redeem(address usr, uint currencyAmount, uint tokenAmount) public;
+    function tokenSupply() public returns (uint);
+}
+
+contract AssessorLike {
+    function calcAndUpdateTokenPrice(address tranche) public returns(uint);
+    function supplyApprove(address tranche, uint currencyAmount) public returns(bool);
+    function redeemApprove(address tranche, uint currencyAmount) public returns(bool);
+    function tokenAmountForONE() public returns(uint);
+}
+
+contract DistributorLike {
+    function balance() public;
+}
+
+contract ProportionalOperator is Math, DSNote, Auth  {
+    TrancheLike public tranche;
+    AssessorLike public assessor;
+    DistributorLike public distributor;
+
+    // lender mappings
+    // each value in a own map for gas-optimization
+    mapping (address => uint) public supplyMaximum;
+    mapping (address => uint) public currentSupplyLimit;
+
+    // expressed in totalCurrencyReturned notation
+    mapping (address => uint) public currencyRedeemed;
+
+    // expressed in totalPrincipalReturned notation
+    mapping (address => uint) public principalRedeemed;
+
+    bool public supplyAllowed  = true;
+
+    uint public totalCurrencyReturned;
+    uint public totalPrincipalReturned;
+
+    // total invested amount of currency and total supply of token
+    // contract implements a fixed supply token price of ONE
+    uint public totalTrancheVolume;
+
+    constructor(address tranche_, address assessor_, address distributor_) public {
+        wards[msg.sender] = 1;
+        tranche = TrancheLike(tranche_);
+        assessor = AssessorLike(assessor_);
+        distributor = DistributorLike(distributor_);
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) public auth {
+        if (contractName == "tranche") { tranche = TrancheLike(addr); }
+        else if (contractName == "assessor") { assessor = AssessorLike(addr); }
+        else if (contractName == "distributor") { distributor = DistributorLike(addr); }
+        else revert();
+    }
+
+    function file(bytes32 what, bool supplyAllowed_) public auth {
+        if(what == "supplyAllowed") {
+            supplyAllowed = supplyAllowed_;
+        }
+    }
+
+    function updateReturned(uint currencyReturned, uint principalReturned) public auth {
+        totalCurrencyReturned += currencyReturned;
+        totalPrincipalReturned += principalReturned;
+    }
+
+    /// defines the max amount of currency for supply
+    function approve(address usr, uint currencyAmount) external auth {
+        supplyMaximum[usr] = currencyAmount;
+        currentSupplyLimit[usr] = currencyAmount;
+    }
```

Comment:
Problem: by overriding both values, we are loosing the entire investment history of the investor that is used in calcMaxRedeem. So if an investor has already supplied some DAI and his allowance is adjusted after, he won't be able to redeem his share as `supplyMaximum[usr] - currentSupplyLimit[usr]` will be 0. 
Possible Solutions: 
` currentSupplyLimit[usr] = currencyAmount - (supplyMaximum[usr] - currentSupplyLimit[usr])`
currencyLimit[usr] has to be an int which is not ideal, since it can turn into a negative value when `currencyAmount is 0 and investedAmount > 0`
so another possible solution would be having 2 different maps:
`currentSupplyLimit[usr]` which stores the currentSupplyLimit of the investor
`redeemedTokens[usr]` which stores the share in the pool that the user received --> can be used for calcMaxRedeem

Reply:
simplified it in the lastest commit